### PR TITLE
Better error messages for invalid literals

### PR DESCRIPTION
**Pull request by Shane32**

### CODE REVIEW

File: src/GraphQL/Validation/Errors.Custom/InvalidLiteralError.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
+using System;
+using GraphQLParser.AST;
+
+namespace GraphQL.Validation
```
Comment (sungam3r) [inline src/GraphQL/Validation/Errors.Custom/InvalidLiteralError.cs:4]: file-scoped

File: src/GraphQL/Execution/ExecutionHelper.cs [hunk 2]
```diff
@@ -52,20 +54,60 @@ public static class ExecutionHelper
...
+                    Directive = directive,
+                    ParentNode = fieldOrFragmentSpread,
+                    Variables = variables,
+                }, arg.DefaultValue);
             }
 
             return values;
         }
 
+        internal readonly ref struct CoerceValueContext
```
Comment (sungam3r) [inline src/GraphQL/Execution/ExecutionHelper.cs:84]: why ref ?

Comment (Shane32) [reply src/GraphQL/Execution/ExecutionHelper.cs:84]: More efficient

Comment (sungam3r) [reply src/GraphQL/Execution/ExecutionHelper.cs:84]: Is it noticeable here in practice or made for general reasons?

Comment (Shane32) [reply src/GraphQL/Execution/ExecutionHelper.cs:84]: ref struct only needs to pass a single pointer to the stack where the struct is stored. Readonly ensures it cannot be changed. Struct would need to assign/copy all variables for each call. So for any readonly struct, readonly ref struct should always be faster (so long as it does not need to be copied to the heap). More so the larger the struct. I’m sure it makes little difference in practice, especially here. It’s internal for now; perhaps if it was public then we would use a regular struct or a class. We also could not use init properties if it was public

File: src/GraphQL.ApiTests/net5/GraphQL.approved.txt [hunk 3]
```diff
@@ -1232,7 +1232,7 @@ namespace GraphQL.Execution
     {
         System.Collections.Generic.IReadOnlyDictionary<GraphQLParser.AST.GraphQLField, System.Collections.Generic.IDictionary<string, GraphQL.Execution.ArgumentValue>>? ArgumentValues { get; }
         System.Threading.CancellationToken CancellationToken { get; }
-        System.Collections.Generic.IReadOnlyDictionary<GraphQLParser.AST.GraphQLField, System.Collections.Generic.IDictionary<string, GraphQL.Execution.DirectiveInfo>>? DirectiveValues { get; }
+        System.Collections.Generic.IReadOnlyDictionary<GraphQLParser.AST.ASTNode, System.Collections.Generic.IDictionary<string, GraphQL.Execution.DirectiveInfo>>? DirectiveValues { get; }
```
Comment (Shane32) [inline src/GraphQL.ApiTests/net5/GraphQL.approved.txt:1235]: `ASTNode` or `IHasDirectivesNode` here?

Comment (sungam3r) [reply src/GraphQL.ApiTests/net5/GraphQL.approved.txt:1235]: `IHasDirectivesNode` is suitable for those who wants to call `.Directives` on it. I don't think someone wants to do it here. I would use `ASTNode`.It looks more logical here.

Comment (Shane32) [reply src/GraphQL.ApiTests/net5/GraphQL.approved.txt:1235]: Makes sense, thanks!

**Comment by Shane32** [comment]
Waiting on #3499

**Comment by sungam3r** [comment]
#3499 has been reviewed

**Comment by sungam3r** [comment]
🤔 
изображение

**Comment by Shane32** [comment]
@sungam3r I think this is ready for review, however, I want to add some more tests to trigger these various literal parsing errors. Feel free to review now, or wait until I remove [WIP] and request a review from you at that time.

**Comment by Shane32** [comment]
I want to continue this work -- that is, my overall goal of allowing validation rules to examine variables easily -- in a subsequent PR.  There may be 2 or 3 more PRs before I'm done.  My work has been stalled because this PR has been marked as WIP when actually I think it's ready for review.  I believe all the changes in this PR are positive; even though further work could be done, such as consolidating error messages between validation code and variable parsing code.  I'll be working to improve all this logic in the upcoming PRs.  But mostly I feel that I can't continue without this being merged.

File: src/GraphQL/Validation/Errors.Custom/InvalidLiteralError.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    /// </summary>
+    public InvalidLiteralError(GraphQLDocument document, ASTNode parentNode, GraphQLDirective? directive, GraphQLArgument? argument, ASTNode node, string message, Exception? innerException = null)
+        : base(document.Source, NUMBER, $"Invalid literal for {GetFieldDescription(parentNode)}{(directive != null ? $" directive '{directive.Name.StringValue}'" : "")}{(argument != null ? $" argument '{argument.Name.StringValue}'" : "")}. {message}", innerException, node)
+    {
+        Code = "INVALID_LITERAL";
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="InvalidLiteralError"/> class for a specified literal
+    /// and exeption. Loads any exception data from the inner exception into this instance.
```
Comment (sungam3r) [inline src/GraphQL/Validation/Errors.Custom/InvalidLiteralError.cs:24]:
```suggestion
    /// and exception. Loads any exception data from the inner exception into this instance.
```

Does this comment also apply to prev ctor?

File: src/GraphQL.Tests/Execution/VariablesTests.cs [hunk 5]
```diff
@@ -187,8 +187,7 @@ public void fail_on_incorrect_value()
             """;
 
         var result = AssertQueryWithErrors(query, null, rules: Enumerable.Empty<IValidationRule>(), expectedErrorCount: 1, executed: false);
-        result.Errors[0].Message.ShouldBe("Error trying to resolve field 'fieldWithObjectInput'.");
-        result.Errors[0].InnerException.Message.ShouldStartWith("Expected object value for 'TestInputObject', found not an object '[\"foo\", \"bar\", \"baz\"]'.");
+        result.Errors[0].Message.ShouldBe("""Invalid literal for field 'fieldWithObjectInput' argument 'input'. Expected object value for 'TestInputObject', found not an object '["foo", "bar", "baz"]'.""");
```
Comment (sungam3r) [inline src/GraphQL.Tests/Execution/VariablesTests.cs:190]: Invalid literal for field 'fieldWithObjectInput' argument 'input'.
->
Invalid literal for argument 'input' of field 'fieldWithObjectInput'.

File: src/GraphQL/Execution/ExecutionHelper.cs [hunk 6]
```diff
@@ -52,20 +54,60 @@ public static class ExecutionHelper
...
+        {
+            public GraphQLDocument? Document { get; init; }
+            /// <summary>
+            /// This is typically either a field, a fragment spread, or variable definition
```
Comment (sungam3r) [inline src/GraphQL/Execution/ExecutionHelper.cs:88]:
```suggestion
            /// This is typically either a field, a fragment spread, or variable definition.
```

File: src/GraphQL/Validation/Errors.Custom/InvalidLiteralError.cs (hunk 4 above)
Comment (Shane32) [reply src/GraphQL/Validation/Errors.Custom/InvalidLiteralError.cs:24]: Fixed.  Yes.  Added text to prev ctor.

File: src/GraphQL.Tests/Execution/VariablesTests.cs (hunk 5 above)
Comment (Shane32) [reply src/GraphQL.Tests/Execution/VariablesTests.cs:190]: Updated throughout.

File: src/GraphQL/Execution/ExecutionHelper.cs (hunk 6 above)
Comment (Shane32) [reply src/GraphQL/Execution/ExecutionHelper.cs:88]: Fixed

**Pull request merged**